### PR TITLE
Closes #6599 validate image source before saving it into the DB

### LINKED ISSUE #6599: Guard beacon script against saving not expected values into the database
**Before submitting an issue please check that you’ve completed the following steps:**
- Made sure you’re on the latest version
- Used the search feature to ensure that the bug hasn’t been reported before

**Describe the bug**
In specific conditions it's possible that our beacon will add unexpected data to the db:
```
{"type":"img","src":"linear-gradient(160deg, rgb(255, 255, 255) 0%, rgb(248, 246, 243) 100%)"}
```
```
  {
    "type": "img",
    "src": "chrome-extension://eanggfilgoajaocelnaflolkadkeghjp/img/commands/main.svg",
    "srcset": "",
    "sizes": "",
    "sources": [],
    "bg_set": [],
    "current_src": "chrome-extension://eanggfilgoajaocelnaflolkadkeghjp/img/commands/main.svg",
    "label": "above-the-fold"
  }
```

**Expected behavior**
We should guard the src against incorrect values. Accepted values are:
1. Absolute URLs of images, starting with `http`, `https`, possibly with `//`
2. Relative URLs of images

**Acceptance Criteria (for WP Media team use only)**
1. Unexpected values not added to the db
2. Images starting with  `http`, `https`, `//` can be added to the database correctly
3. Relative URLs of images added correctly to the db
4. No regressions in handling different type of implemented markups (background, responsive, picture, basic images)

### PR DESCRIPTION
**Pull request by wordpressfan**
# Description

Fixes #6599

### User documentation

We just add another layer of sanitizing and checking the valid lcp/atf to be valid images

### Technical documentation

Here we will exclude the following patterns and more:

```
https://domain.ext/file.php?url=img.jpg
https://domain.ext/file.js?url=img.jpg
https://domain.ext/file.php#url=img.jpg
chrome-extension://extension-hash/path/to/image/x.svg
linear-gradient(160deg, rgb(255, 255, 255) 0%, rgb(248, 246, 243) 100%)
```

## Type of change

- [x] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
None

## Risks
None

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [ ] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.
- [ ] I listed the introduced external dependencies explicitely on the PR.
- [ ] I validated the repo-specific guidelines from CONTRIBUTING.md.

## Observability
- [x] I handled errors when needed.
- [x] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

### CODE REVIEW

**Comment by piotrbak** [comment]
Could we also validate the resource usage/time spent on this one?

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: suggesting for readability to assign the value of the method to variable `$valid_image`, and use the variable into the filter

File: tests/Unit/inc/Engine/Media/AboveTheFold/AJAX/Controller/addLcpData.php
Comment (remyperona) [inline tests/Unit/inc/Engine/Media/AboveTheFold/AJAX/Controller/addLcpData.php]: We have a stub function `stubWpParseUrl()` that handles this

**Review by remyperona — CHANGES_REQUESTED** [review]

File: tests/Unit/inc/Engine/Media/AboveTheFold/AJAX/Controller/addLcpData.php
Comment (Miraeld) [reply tests/Unit/inc/Engine/Media/AboveTheFold/AJAX/Controller/addLcpData.php]: Fixed :)

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (Miraeld) [reply inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: Fixed :)

**Review by Miraeld — APPROVED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Mai-Saad** [comment]
Reproducible for links here except for linear gradient case ... now we aren't adding those kind of URLs to DB. However, on this template, https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp .webp is atf and not LCP while in PSI it's LCP https://e2e.rocketlabsqa.ovh/lcp_6599_template2/ (needs further investigation)

Note: on trunk, webP is set as lcp (while we commented the BG image code but it isn't on branch)

**Comment by jeawhanlee** [comment]
@wordpressfan  We have a regression here while running e2e tests, see [here](https://wp-media.slack.com/archives/CUT7FLHF1/p1717427420573809?thread_ts=1717417271.741509&cid=CUT7FLHF1)

**Comment by MathieuLamiot** [comment]
Manual tests 🆗  https://wp-media.slack.com/archives/CUKB44GNN/p1717491699283409?thread_ts=1717423732.490199&cid=CUKB44GNN
Rocket-E2E KO, need to retry after fixing #6693 (which is the root cause of the aforementioned regression) 
Moving to blocked

**Comment by Mai-Saad** [comment]
The test template is fine after merging 6693 to the branch, lcp not found and webP is ATF https://new.rocketlabsqa.ovh/lcp_6599_template2/.

Note: visit the mobile version of the page using browserstack, is not added to DB (same on the trunk), visiting normal WP mobile page added normally to ATF table ===> In general further investigation is needed to make the template works

If e2e is green then we can merge

**Comment by jeawhanlee** [comment]
E2E Test Report: https://wp-media.slack.com/docs/T02RYMVKQ/F076AFQK40P

**Comment by MathieuLamiot** [comment]
@wordpressfan Is that expected that images from another hostname are filtered out with this PR?
If yes, then we have a lot of test failing because of this and a lot of clean up will be necessary on Rocket-E2E.
If not, then the report must be further investigated.

@wordpressfan There seems to be a regression, for instance on the template `http://mike.e2e.rocketlabsqa.ovh/lcp_no_dimensions_picture`, the LCP is 'not found' instead of `/wp-content/rocket-test-data/images/lcp/testavif.avif` and I don't think it is expected? Note that this image testavif.avif is filtered out on many templates, and it looks like its PHP that filters it out. I don't think this is expected. Can you check? Seems to happen with webp too sometimes (`testwebp.webp`)

@piotrbak I would advice not to move forward with this PR in 3.16.1 as the result is quite uncertain at this stage.

**Comment by wordpressfan** [comment]
We changed the structure of the data being saved into the database a bit in other PRs and this PR is validating based on the old structure, so for sure many things will fail and we need to check, that's why this unit test case is failing as an example.

**Comment by wordpressfan** [comment]
I need the product decision about the following case:

We were bailing out when the image src attribute is not valid, now we support having multiple background images (bg_set) so we need also to check those background image urls one by one and remove the not valid ones not to be saved in the database? or just we need to check the src if found?

@wp-media/productrocket

**Comment by MathieuLamiot** [comment]
Product answer: https://wp-media.slack.com/archives/CUT7FLHF1/p1717696364743869?thread_ts=1717695053.640209&cid=CUT7FLHF1

So now that the CI passes, we'll need to run Rocket-E2E, if OK move to Ready for QA, and from here, manually test.

**Comment by jeawhanlee** [comment]
E2E test fails for a single template ( `lcp_no_dimension_svg` ) on desktop

`Expected LCP for desktop - /wp-content/rocket-test-data/images/lcp/testsvg.svg for http://mike.e2e.rocketabsqa.ovh/lcp_no_dimension_svg is not present in actual - not found`

On Mobile
`Expected LCP for mobile - https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp for http://mike.e2e.rocketlabsqa.ovh/lcp_with_space_after_title is not present in actual - not found`

`Expected Viewport for mobile - https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/test_inline2.jpeg for http://mike.e2e.rocketlabsqa.ovh/lcp_with_space_after_title is not present in actual - []`

By the way, we have the same failure for mobile on develop as well

**Comment by MathieuLamiot** [comment]
Thanks a lot @jeawhanlee 🙏 
Pending Product direction for svg: https://wp-media.slack.com/archives/CUT7FLHF1/p1717765436734749?thread_ts=1717695053.640209&cid=CUT7FLHF1

**Comment by MathieuLamiot** [comment]
@jeawhanlee As per product direction [here](https://wp-media.slack.com/archives/CUT7FLHF1/p1717769925268999?thread_ts=1717695053.640209&cid=CUT7FLHF1): we must not exclude svg images. So the failing E2E test must be fixed by changing the wp-rocket code. This PR is currently introducing a regression.

My gut feeling is that svg might not be supported by `wp_check_filetype` added in this PR? (Just a gut feeling, I did not check anything). To be investigated and fixed.

Note that, in the AC, we should exclude `chrome-extension://extension-hash/path/to/image/x.svg` (which is done, according to the integration test added in the PR), but we should because of the protocol, not the image type.
Maybe adding an integration test with a "normal" https svg image would allow us to refine.

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (wordpressfan) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: This will add the support of svg in the whole site not that specific feature, do we really want to do that or we just need it to be in our feature only?

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (wordpressfan) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: That's a good idea to bailout early when finding this chrome scheme but what do u think if we parsed the url to get the scheme also and we can exclude whatever scheme we want, because we may increase this list of exclusions in the future, so what do u think if we use `wp_parse_url` to get also the scheme and exclude the scheme `chrome-extension`
So if we wanted to increase the list we can make it an array of exclusions and have a filter.

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (MathieuLamiot) [reply inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: Just for our feature. I agree we should not touch the mime_types filter globally

**Comment by jeawhanlee** [comment]
Ran e2e tests 2x and we are good for desktop, mobile scenario failures still the same as the one [here](https://github.com/wp-media/wp-rocket/pull/6605#issuecomment-2154780167)

File: inc/Engine/Media/AboveTheFold/AJAX/Controller.php
Comment (wordpressfan) [inline inc/Engine/Media/AboveTheFold/AJAX/Controller.php]:
```suggestion
		/**
		 * Filters the supported schemes of LCP/ATF images.
		 *
		 * @param array  $invalid_schemes Array of invalid schemes.
		 */
		$invalid_schemes = apply_filters( 'rocket_atf_invalid_schemes', array( 'chrome-extension' ) );

		$image_src_components = wp_parse_url( $image_src );

		if ( empty( $image_src_components['path'] ) || ( ! empty( $image_src_components['scheme'] ) && in_array( strtolower( $image_src_components['scheme'] ), $invalid_schemes, true ) ) ) {
			return false;
		}

		// Add svg to allowed mime types.
		$allowed_mime_types        = get_allowed_mime_types();
		$allowed_mime_types['svg'] = 'image/svg+xml';
		$image_src_filetype_array  = wp_check_filetype( $image_src_components['path'], $allowed_mime_types );
```

This is what I meant, what do u think?

Comment (jeawhanlee) [reply inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: @wordpressfan This is fine, but I prefer the flexibility that comes with regex as we could have `chrome-extensions:// `but we could also have `chrome-untrusted://` in this case, we would have to try to capture all kinds of `chrome-*` in the list, this is the reason I went this way.

Comment (wordpressfan) [reply inc/Engine/Media/AboveTheFold/AJAX/Controller.php]: As you like, I just hate regex :D because it can throw unexpected failures regarding some inputs or it may take time more than in_array but I'm sure it'll do the trick.

Can you plz just guard the filter return to be an array and not empty?

**Comment by wordpressfan** [comment]
I think it's working now properly, thanks @jeawhanlee 
I can't approve it because the PR is mine

**Comment by jeawhanlee** [comment]
I can't approve too because technically it's mine as well 😁

**Review by remyperona — CHANGES_REQUESTED** [review]
It's missing the minified version of the beacon script

**Comment by remyperona** [comment]
Can you update the PR description to remove irrelevant data and complete the checklist?

**Comment by jeawhanlee** [comment]
I have run e2e on this branch again after merging develop into it and I get just one template failing on desktop but this template is also failing on dev, so no regression against dev here:

`Expected Viewport for desktop - wp-content/rocket-test-data/images/lcp/testjpg.jpg for http://mike.e2e.rocketlabsqa.ovh/lcp_no_dimensions_picture is not present in actual - [{"type":"picture","src":"http://mike.e2e.rocketlabsqa.ovh/test.png","sources":[]},{"type":"picture","src":"http://mike.e2e.rocketlabsqa.ovh/wp-content/rocket-test-data/images/test3.gif","sources":[{"srcset":"http://mike.e2e.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp","media":"(max-width: 700px)","type":"","sizes":""},{"srcset":"http://mike.e2e.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testjpeg.jpeg","media":"(max-width: 450px)","type":"","sizes":""}]}]`

**Review by Miraeld — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
Next step to test this PR is to re-run E2E once https://github.com/wp-media/wp-rocket-e2e/issues/88 is implemented.

**Comment by MathieuLamiot** [comment]
E2E Results with E2E commit 7edc2237af34f62bf79185baa39023c8d83e0afd:

```
Expected Viewport for desktop - https://new.rocketlabsqa.ovh/file.php?url=img.jpg for http://mathieu.e2e.rocketlabsqa.ovh/lcp_6599_template2 is not present in actual - [{"type":"img","src":"https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp"}]

Expected Viewport for mobile - https://new.rocketlabsqa.ovh/file.php?url=img.jpg for http://mathieu.e2e.rocketlabsqa.ovh/lcp_6599_template2 is not present in actual - [{"type":"img","src":"https://new.rocketlabsqa.ovh/wp-content/rocket-test-data/images/lcp/testwebp.webp"}]
```

According to the expected behavior [here](https://github.com/wp-media/wp-rocket/issues/6599#issuecomment-2080947377), I think it's OK and that the E2E must be adapted. Here is my suggestion for the PR to adapt: https://github.com/wp-media/wp-rocket-e2e/pull/117

Tested with the E2E PR, it's all OK.

I'll move this to QA Done, and wait for QA team to come back to merge.

**Pull request merged**
